### ISSUE TITLE
Support for Map of parametrized type

### ISSUE THREAD

**Issue by TehBakker** [bolcom/spring-data-mongodb-encrypt#9]
Greetings,

Nice library but sadly it doesnt seems to work for fields that are of type Map<String, List<T>> 

When I add the @Field on the map attribute, then on the T object sub-document I get an error on startup 

> org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'encryptionEventListener': Invocation of init method failed; nested exception is java.lang.IllegalArgumentException: mypackage.api.server.dao.collections.onboarding.OnboardingItineraryEntity.listSessions
> 	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:137) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.applyBeanPostProcessorsBeforeInitialization(AbstractAutowireCapableBeanFactory.java:409) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1620) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:555) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:483) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:761) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:867) ~[spring-context-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:543) ~[spring-context-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:122) ~[spring-boot-1.5.6.RELEASE.jar:1.5.6.RELEASE]
> 	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:693) ~[spring-boot-1.5.6.RELEASE.jar:1.5.6.RELEASE]
> 	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:360) ~[spring-boot-1.5.6.RELEASE.jar:1.5.6.RELEASE]
> 	at org.springframework.boot.SpringApplication.run(SpringApplication.java:303) ~[spring-boot-1.5.6.RELEASE.jar:1.5.6.RELEASE]
> 	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:134) [spring-boot-1.5.6.RELEASE.jar:1.5.6.RELEASE]
> 	at mypackage.api.server.Application.main(Application.java:32) [classes/:na]
> Caused by: java.lang.IllegalArgumentException: mypackage.api.server.dao.collections.onboarding.OnboardingItineraryEntity.listSessions
> 	at com.bol.reflection.ReflectionCache.processDocument(ReflectionCache.java:67) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	at com.bol.reflection.ReflectionCache.processDocument(ReflectionCache.java:19) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	at com.bol.secure.EncryptionEventListener.lambda$initReflection$0(EncryptionEventListener.java:41) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	at java.lang.Iterable.forEach(Iterable.java:75) ~[na:1.8.0_111]
> 	at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080) ~[na:1.8.0_111]
> 	at com.bol.secure.EncryptionEventListener.initReflection(EncryptionEventListener.java:40) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_111]
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_111]
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_111]
> 	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_111]
> 	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleElement.invoke(InitDestroyAnnotationBeanPostProcessor.java:366) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor$LifecycleMetadata.invokeInitMethods(InitDestroyAnnotationBeanPostProcessor.java:311) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	at org.springframework.beans.factory.annotation.InitDestroyAnnotationBeanPostProcessor.postProcessBeforeInitialization(InitDestroyAnnotationBeanPostProcessor.java:134) ~[spring-beans-4.3.10.RELEASE.jar:4.3.10.RELEASE]
> 	... 17 common frames omitted
> Caused by: java.lang.IllegalArgumentException: mypackage.api.server.dao.collections.onboarding.OnboardingSessionEntity.mapAnalysis
> 	at com.bol.reflection.ReflectionCache.processDocument(ReflectionCache.java:67) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	at com.bol.reflection.ReflectionCache.processDocument(ReflectionCache.java:46) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	... 29 common frames omitted
> Caused by: java.lang.ClassCastException: sun.reflect.generics.reflectiveObjects.ParameterizedTypeImpl cannot be cast to java.lang.Class
> 	at com.bol.reflection.ReflectionCache.processDocument(ReflectionCache.java:52) ~[spring-data-mongodb-encrypt-1.0.1.jar:na]
> 	... 30 common frames omitted

**Comment by agoston** [comment]
This use case, embedding Collections in each other, is indeed not tested or used before.
Will look into this.

**Comment by agoston** [comment]
Hi!

Apparently, this does not occur that simply. I've added an example test and was surprised to see it works:

https://github.com/bolcom/spring-data-mongodb-encrypt/blob/master/src/test/java/com/bol/system/MyBean.java

I think you need a `Map<String, Map<>>` to actually trigger this.

**Comment by TehBakker** [comment]
Yep the @Encrypted was on sub fields.
I had a Map<String, List<MyObj>>
With MyObj having an @Encrypted field.

**Comment by agoston** [comment]
I just managed to successfully cover this scenario:

```
    @Field
    public Set<Map<String, List<Set<Map<String, MyBean>>>>> nestedCollectionsBean;
```

I will add system test and make a release.

**Comment by agoston** [comment]
Released 1.2.0. Can you give it a try?

**Comment by TehBakker** [comment]
Hey,
Sorry changed my model from Map<String, List<Obj>> to List<Obj> so can't really retry right away.
Thanks for the quick fix though !
